pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +294,171 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
+	switch len(parts) {
+	case 1:
+		name = parts[0]
+	case 2:
+		var prefix string
+		prefix, name = parts[0], parts[1]
+		if len(prefix) == 0 {
+			errs = append(errs, "prefix of the topology key must be non-empty")
+			//} else if msgs := validation.IsDNS1035Label(prefix); len(msgs) != 0 {
+		} else if !topologyPrefixRexp.MatchString(prefix) {
+			errs = append(errs, "prefix of the topology key "+validation.RegexError(topologyPrefixRexpErrMsg, topologyPrefixRexpFmt, "company.com"))
+		}
+	default:
+		return append(errs, validTopologyKeySegmentErrMsg)
+	}
+
+	if len(name) == 0 {
+		errs = append(errs, "name of the topology key must be non-empty")
+	} else if len(name) > topologyNameMaxLength {
+		errs = append(errs, "name of the topology key must be 63 characters or less.")
+	}
+	if !topologyNameRexp.MatchString(name) {
+		errs = append(errs, "name of the topology key "+validation.RegexError(topologyNameRexpErrMsg, topologyNameRexpFmt, "csi-driver"))
+	}
+	return errs
+}
+
+// ValidateCSIDriver validates a CSIDriver.
+func ValidateCSIDriver(csiDriver *storage.CSIDriver) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, apivalidation.ValidateCSIDriverName(csiDriver.Name, field.NewPath("name"))...)
+
+	allErrs = append(allErrs, validateCSIDriverSpec(&csiDriver.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateCSIDriverUpdate validates a CSIDriver.
+func ValidateCSIDriverUpdate(new, old *storage.CSIDriver) field.ErrorList {
+	allErrs := ValidateCSIDriver(new)
+
+	// Spec is read-only
+	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
+	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), new.Spec, "field is immutable"))
+	}
+	return allErrs
+}
+
+// ValidateCSIDriverSpec tests that the specified CSIDriverSpec
+// has valid data.
+func validateCSIDriverSpec(
+	spec *storage.CSIDriverSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validateAttachRequired(spec.AttachRequired, fldPath.Child("attachedRequired"))...)
+	allErrs = append(allErrs, validatePodInfoOnMountVersion(spec.PodInfoOnMountVersion, fldPath.Child("podInfoOnMountVersion"))...)
+	return allErrs
+}
+
+// validateAttachRequired tests if attachRequired is set for CSIDriver.
+func validateAttachRequired(attachRequired *bool, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if attachRequired == nil {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	}
+
+	return allErrs
+}
+
+// validatePodInfoOnMountVersion tests if the podInfoOnMountVersion is valid for CSIDriver.
+func validatePodInfoOnMountVersion(podInfoOnMountVersion *string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if podInfoOnMountVersion != nil && len(*podInfoOnMountVersion) > 0 && !supportedPodInfoOnMountVersions.Has(string(*podInfoOnMountVersion)) {
```

Comment:
do we really want to allow empty string? Can we just have nil as the "unspecified" value?

Reply:
Suppose user has PodInfoOnMountVersion in a yaml file, I think it makes sense to allow it to be empty which means it is not set, same as if PodInfoOnMountVersion is not in the yaml file.

Reply:
The point of making it a pointer is to tell when it is unset.  There's anti-value in using "" to mean unset - patches become ambiguous.

Reply:
Ok.  Disallow empty string now.